README
Forked from https://bitbucket.org/speedbird/pyforum.

HISTORY
The "Release Number" is located at the bottom-right of the UI

Release 1.1.3
================
- Forked from the bitbucket repo Nov 29, 2012
- Various fixes and improvements.

Release 1.1.2
================
- ????

Release 1.1 Beta
================
- Released TBD
- Revamped core (security assertions) to address perceived secutiry flaws
  by some entities (credible and otherwise)
- Removed old google authentication and replaced with multiple sign-in
  options (Janrain).

Release 1.0.3
=============
- Released Oct 13, 2009
- Anonymous postings now require captcha-like authorization for replying to posts
- Few CSS changes to make the UI a little bit more consistent

Release 1.0.2
=============
- Released Sep 13, 2009
- Authentication via Google Account login enabled
- Style Sheet overhaul, smaller font size overall
- New Logo

Release 1.0.1
=============
- Released May 27, 2009
- Fixed several UI elements (font sizes and such)
- Changed Logo (again) for a more subtle one.

Release 1.0.0
=============
- Released May 25, 2009
- Require "captcha"-style humanity verification challenge when sending messages as anonymous users or
  when forum administrators allow anonymous users to add/reply to posts.
- Fixed bug in Forum Administration when setting up access properties for a forum.
- "Administrator" user is automatically created when the forum system is installed for the first time
- New application logo.
- Added "Purge" functionality to allow remove old administrator messages

Release Candidate 4
===================
- Released May 13, 2009
- Added CSS class to properly handle code snippets
- replaced soon-to-be-deprecated python library sha in favor of hashlib
- Administration - User list now displays join in date and table columns are sortable
- Spanish language enabled
- German language enabled
- Allow contact users via PM system can be now triggered properly via user's preferences

Release Candidate 3
===================
- Released April 22, 2009
- Fixed bug that allowed anonymous users access to (their "own") personal messages
- Removed links in member names when forum is viewed by anonymous users
- Administrators can change the roles of any users from the user edit page

Release Candidate 2
===================
- Released April 19, 2009
- Changed Distribution License from ZPL 2.1 to GPL 3.0
- Changed footer information to allow modification by users for other parts of the program
- Changed several references to "zforum" (original name) to "pyforum", including several CSS classes

Release Candidate 1
===================
- Released April 18, 2008
- Personal Messages functionality completed
- Broken image placeholder for users without avatar selected fixed

Beta Release 6
==============
- Released Apr 16, 2009
- Fixed RSS feeds, broken since Beta 3
- Added partial PM functionality, users still cannot pm other users

Beta Release 5
==============
- Released Apr 15, 2009
- Added Pagination widget
- Fixed "Reply w/Quote" capabilities

Beta release 4
==============
- Released Apr 14 2009
- Enhanced Emoticon handling
- Added pseudo-html code and parsing capabilities
- Added "Reply w/Quote" capabilities (introduced a bug)

Beta Release 3
==============
- Released Apr 13 2009
- Added Emoticons

Beta Release 2
==============
- Released Apr 11 2009
- Bug Fixes

Beta Release 1
==============
- Released April 7, 2009
- Initial Release


models/parser.py
# Provides a secure html/xhtml parser for the system
# Use from your controllers:
# text = parse_content(text)

import sgmllib, string
from gluon.html import URL
import re

class StrippingParser(sgmllib.SGMLParser):
    # These are the HTML tags that we will leave intact
    valid_tags = ('b', 'i', 'u', 'code',)
    tolerate_missing_closing_tags = ('br',)
    from htmlentitydefs import entitydefs # replace entitydefs from sgmllib

    def __init__(self, mode):
        sgmllib.SGMLParser.__init__(self)
        self.result = []
        self.endTagList = []
        self.mode = mode

    def handle_data(self, data):
        self.result.append(data)

    def handle_charref(self, name):
        self.result.append("&#%s;" % name)

    def handle_entityref(self, name):
        x = ';' * self.entitydefs.has_key(name)
        self.result.append("&%s%s" % (name, x))

    def unknown_starttag(self, tag, attrs):
        """ Delete all tags except for legal ones. """
        if self.mode == "removeall":
            if tag in self.valid_tags:
                self.result.append('<' + tag)
                for k, v in attrs:
                    if string.lower(k[0:2]) != 'on' and string.lower(v[0:10]) != 'javascript':
                        self.result.append(' %s="%s"' % (k, v))
                self.result.append('>')
                if tag not in self.tolerate_missing_closing_tags:
                    endTag = '</%s>' % tag
                    self.endTagList.insert(0,endTag)

    def unknown_endtag(self, tag):
        if self.mode == "removeall":
            if tag in self.valid_tags:
                # We don't ensure proper nesting of opening/closing tags
                endTag = '</%s>' % tag
                self.result.append(endTag)
                self.endTagList.remove(endTag)

    def cleanup(self):
        """ Append missing closing tags. """
        self.result.extend(self.endTagList)

    def get_html(self):
        html = ' '.join(self.result)
        #html = html.replace('<br>', '<br />')
        #html = html.replace('\r\n', '\n') # Windows
        #html = html.replace('\n', '<br />')
        return html

    def parse_emoticons(self, html_code):
        """ This is pyForum Specific """
        # Emoticon Handling
        emoticons = ['icon_arrow.png', 'icon_biggrin.png', 'icon_confused.png', 'icon_cool.png', 'icon_cry.png', 'icon_exclaim.png', 'icon_idea.png', 'icon_lol.png', 'icon_mad.png', 'icon_mrgreen.png', 'icon_neutral.png', 'icon_question.png', 'icon_razz.png', 'icon_redface.png', 'icon_rolleyes.png', 'icon_sad.png', 'icon_smile.png', 'icon_twisted.png', 'icon_wink.png']
        for icon in emoticons:
            replace_tag = (icon[5:icon.rfind('.')])
            replace_tag_with = '<img src="%s" alt="%s" style="width:16px;height:16px;" class="imgclear" style="vertical-align:middle;" />' % (URL(r=request, c='static/images', f=icon), replace_tag)
            html_code = html_code.replace(':%s:' % (replace_tag), replace_tag_with)
        return html_code
        
    def parse_pseudo_html(self, html_code):
        """ This is also pyForum Specific """
        tags = {'[b]': '<b>',
                '[/b]': '</b>',
                '[u]': '<u>',
                '[/u]': '</u>',
                '[i]': '<i>',
                '[/i]': '</i>',
                '[code]': '<pre class="code">',
                '[/code]': '</pre>',
                '[quote]': '<div class="quote"><div>',
                '[/quote]': '</div></div>',
                '[small]': '<span class="small">',
                '[/small]': '</span>',
                '[smallb]': '<span class="smallb">',
                '[/smallb]': '</span>'}
        pseudo_html = self.multiple_replace(tags, html_code)
        # Now on to URLs
        pseudo_html = re.sub('(\[url\])(.*?)(\[/url\])', '<a href="\\2" title="\\2">\\2</a>', pseudo_html)
        # Next images
        pseudo_html = re.sub('(\[img\])(.*?)(\[/img\])', '<img src="\\2" alt="\\2" />', pseudo_html)
        # and finally some formatting for the final display
        pseudo_html = pseudo_html.replace('<br>', '<br />')
        pseudo_html = pseudo_html.replace('\r\n', '<br />') # Windows
        pseudo_html = pseudo_html.replace('\n', '<br />')
        return pseudo_html
        
        
    def multiple_replace(self, dict, html_code): 
        """ Replace in 'text' all occurences of any key in the given
        dictionary by its corresponding value.  Returns the new tring.""" 
        # Create a regular expression  from the dictionary keys
        regex = re.compile("(%s)" % "|".join(map(re.escape, dict.keys())))
        # For each match, look-up corresponding value in dictionary
        return regex.sub(lambda mo: dict[mo.string[mo.start():mo.end()]], html_code) 


def parse_content(content, mode="removeall"):
    """ Parse the messages mode can be "removeall", "forumpreview" and "forumfull" """
    if content.strip():
        parser = StrippingParser(mode)
        new_content = content
        if mode == "forumpreview":
            # Here we need to "trick" the parser and "pre-convert" our custom-html
            # code from :pseudocode: into the "real" html, so it is stripped off
            new_content = parser.parse_emoticons(new_content)
            new_content = parser.parse_pseudo_html(new_content)
            
        parser.feed(new_content)
        parser.close()
        parser.cleanup()
        clean_content = parser.get_html()
        if mode == "forumfull":
            # Here we need to "convert" emoticons and images
            clean_content = parser.parse_emoticons(clean_content)
            clean_content = parser.parse_pseudo_html(clean_content)
    else:
        clean_content = ''
    return clean_content
   



modules/forumhelper.py
import base64
import struct
import smtplib
from gluon.html import URL
import random
import hashlib
from cStringIO import StringIO
import math

class ForumHelper(object):
    """ Provides methods to deal with the forum's miscellaneous
    functions """

    def __init__(self, environment, db, auth_user):
        self.request = environment['request']
        self.response = environment['response']
        self.session = environment['session']
        self.cache = environment['cache']
        self.db = db
        self.auth_user = auth_user

    def get_system_property(self, property_name, default_value):
        """ Checks for a possible system property and returns its
        associated value if the property exists and its value is empty,
        it does also return the default value

        """
        prop = self.db(self.db.zf_system_properties.property_name == \
            property_name).select(self.db.zf_system_properties.id,
                                  self.db.zf_system_properties.property_desc,
                                  self.db.zf_system_properties.property_value)
        if prop:
            property_value = prop[0].property_value
        else:
            property_value = default_value
        return property_value

    def get_system_properties(self):
        """ Returns all of pyForum's system properties """
        return self.db().select(self.db.zf_system_properties.ALL)

    def put_system_property(self, property_name, new_value):
        """ UPDATES an existing property name if it does indeed
        exist, "new value" must always be a string,
        it returns a boolean value describing success in adding the property
        or failure when the prioperty could not be updated """
        prop = self.db(self.db.zf_system_properties.property_name == \
            property_name).select(self.db.zf_system_properties.id)
        if prop:
            prop_id = prop[0].id
            self.db(self.db.zf_system_properties.id == \
                    prop_id).update(property_value=new_value)
            updated = True
        else:
            updated = False
        return updated

    def get_member_property(self, property_name, user_id, default_value):
        """ Similar to get_system_property() but will handle member properties
        instead of system properties

        """
        user_property_value = default_value
        # First check if the property exists
        property_check = self.db(
            self.db.zf_member_properties_skel.property_name == \
            property_name).select(self.db.zf_member_properties_skel.id)
        if (property_check):
            property_id = property_check[0].id
            # Now check if the user *has* this property in his profile
            user_prop = self.db((self.db.zf_member_properties.property_id == \
                property_id) & (self.db.zf_member_properties.user_id == \
                user_id)).select(self.db.zf_member_properties.property_value)
            if user_prop:
                user_property_value = user_prop[0].property_value
        return user_property_value

    def get_display_name(self, user_id=None, default=None):
        """ This method is a shortcut for calling get_member_property()
        with the apporpriate parameter. It will return the user's "forum"
        name, which can be anything the user chooses in his/her profile

        """
        if user_id is None:
            user_id = self.auth_user.get_user_id()

        if default is None:
            default = 'user_%s' % (user_id) # Kind of a convention only
        return self.get_member_property('zfmp_display_name', user_id, default)

    def put_member_property(self, property_name, user_id, new_value):
        errors = {'errors': ''}
        # Check if the property exists
        property_check = self.db(
            self.db.zf_member_properties_skel.property_name == \
                property_name).select(
            self.db.zf_member_properties_skel.id,
            self.db.zf_member_properties_skel.member_editable)
        if (property_check):
            property_id = property_check[0].id
            curr_user_id = self.auth_user.get_user_id()
            #member_editable = property_check[0].member_editable
            if curr_user_id == user_id or self.auth_user.is_admin():
                # Does this property exists for the user?
                prop_exists = self.db(
                    (self.db.zf_member_properties.property_id == \
                     property_id) & \
                    (self.db.zf_member_properties.user_id == \
                     user_id)).select(self.db.zf_member_properties.id)
                if (prop_exists):
                    # Yes, then update
                    self.db(
                        (self.db.zf_member_properties.property_id == \
                        property_id) & \
                        (self.db.zf_member_properties.user_id == \
                         user_id)).update(property_value=new_value)
                else:
                    # Nope, Create
                    self.db.zf_member_properties.insert(
                        user_id=user_id,
                        property_id=property_id, property_value=new_value)
            else:
                errors.update({'errors':
                    'Not allowed to change this property'})
        else:
            errors.update({'errors': 'Member property name not found'})
        return errors

    def has_member_avatar(self, user_id, bypass=True):
        """ Tests if there is an avatar stored for the user, if bypass is
        True, it'll check for the avatar active flag

        """
        if bypass:
            avatar_info = self.db(\
                    (self.db.zf_member_avatars.user_id == user_id) &
                    (self.db.zf_member_avatars.avatar_active == \
                     True)).select(self.db.zf_member_avatars.content_type,
                                   self.db.zf_member_avatars.avatar_image)
        else:
            avatar_info = self.db(self.db.zf_member_avatars.user_id == \
                self.db.zf_member_avatars.user_id).select(
                self.db.zf_member_avatars.content_type,
                self.db.zf_member_avatars.avatar_image)
        if (avatar_info):
            return True
        else:
            return False

    def has_forum_subscription(self, forum_id, user_id):
        """ Returns True if auth_user is subscribed to the requested forum,
        False otherwise

        """
        return self._has_subscription(forum_id, 'F', user_id)

    def has_topic_subscription(self, topic_id, user_id):
        """ Returns True if auth_user is subscribed to the requested topic,
        False otherwise

        """
        return self._has_subscription(topic_id, 'T', user_id)

    def _has_subscription(self, object_id, object_type, user_id):
        """ Returns True if auth_user is subscribed to the requested
        forum/topic, False otherwise

        """
        subscribed = False
        #raise ValueError, self.auth_user.is_auth()
        if user_id:
            if self.db(
                (self.db.zf_member_subscriptions.user_id == user_id) &
                (self.db.zf_member_subscriptions.subscription_type == \
                 object_type) &
                (self.db.zf_member_subscriptions.subscription_active == True) &
                (self.db.zf_member_subscriptions.subscription_id == \
                 object_id)).select(self.db.zf_member_subscriptions.id):
                subscribed = True
        return subscribed

    def add_topic_subscription(self, topic_id, user_id):
        """ Handles Topic Subscriptions """
        return self._handle_subscription(topic_id, 'T', user_id, 'add')

    def del_topic_subscription(self, topic_id, user_id):
        """ Handles Topic Subscriptions """
        return self._handle_subscription(topic_id, 'T', user_id, 'remove')

    def add_forum_subscription(self, forum_id, user_id):
        """ Handles Forum Subscriptions """
        return self._handle_subscription(forum_id, 'F', user_id, 'add')

    def del_forum_subscription(self, forum_id, user_id):
        """ Handles Forum Subscriptions """
        return self._handle_subscription(forum_id, 'F', user_id, 'remove')

    def _handle_subscription(self, object_id, object_type, user_id, action):
        """ object_id is a forum_id or a topic_id, object_type is "F" for
        forum, "T" for topic, action is "add" or "remove"

        """
        success = True
        # See if the record actually exists
        subscription = self.db(
            (self.db.zf_member_subscriptions.subscription_id == object_id) &
            (self.db.zf_member_subscriptions.user_id == user_id) &
            (self.db.zf_member_subscriptions.subscription_type == \
             object_type)).select()
        if (subscription):
            # Ok, there is a record, this means that we'll update it rather
            # than add a new one
            if action == "add":
                self.db(
                    (self.db.zf_member_subscriptions.subscription_id == \
                     object_id) &
                    (self.db.zf_member_subscriptions.user_id == user_id) &
                    (self.db.zf_member_subscriptions.subscription_type == \
                     object_type)).update(subscription_active=True)
            elif action == "remove":
                self.db(
                    (self.db.zf_member_subscriptions.subscription_id == \
                     object_id) &
                    (self.db.zf_member_subscriptions.user_id == user_id) &
                    (self.db.zf_member_subscriptions.subscription_type == \
                     object_type)).update(subscription_active=False)
            else:
                success = False
        else:
            # No record found, add a brand new one, in this case there
            # can be no "remove"
            self.db.zf_member_subscriptions.insert(
                user_id=user_id,
                subscription_id=object_id,
                subscription_type=object_type,
                subscription_active=True)
        return success

    def get_user_rank(self, user_id):
        """ Returns the rank of the user """
        rank = ''
        hits = int(self.get_member_property('zfmp_postings', user_id, 0))
        rank_info = self.db(
            self.db.zf_member_rank.rank_value_min <= hits).select(
            self.db.zf_member_rank.rank_name,
            orderby=~self.db.zf_member_rank.rank_value_min, limitby=(0,1))
        #raise ValueError, rank_info
        if (rank_info):
            rank = rank_info[0].rank_name
        return rank

    def get_system_announcements(self, include_content=False, rss=False):
        max_sys_announcements = int(
            self.get_system_property('zfsp_system_announcement_max', 0))
        sys_topics = None
        if max_sys_announcements:
            if include_content:
                sys_topics = self.db(
                    (self.db.zf_topic.system_announcement_flag == True) &
                    (self.db.zf_topic.parent_flag == True)).select(
                    self.db.zf_topic.title,
                    self.db.zf_topic.id,
                    self.db.zf_topic.modifying_date,
                    self.db.zf_topic.content,
                    orderby=~self.db.zf_topic.modifying_date,
                    limitby=(0, max_sys_announcements))
            else:
                sys_topics = self.db(
                    (self.db.zf_topic.system_announcement_flag == True) &
                    (self.db.zf_topic.parent_flag == True)).select(
                    self.db.zf_topic.title,
                    self.db.zf_topic.id,
                    self.db.zf_topic.modifying_date,
                    orderby=~self.db.zf_topic.modifying_date,
                    limitby=(0, max_sys_announcements))
            if not (sys_topics) and not rss:
                sys_topics = [{'error':'No Topics'}]
        else:
            if not rss:
                sys_topics = [{'error':'No Topics'}]
        return sys_topics

    def get_latest_topics(self, include_content=False, rss=False):
        max_topics = int(
            self.get_system_property('zfsp_latest_postings_max', 0))
        latest_topics = None
        if max_topics:
            if include_content:
                latest_topics = self.db(\
                        (self.db.zf_topic.system_announcement_flag == False) &
                        (self.db.zf_forum.include_latest_topics == True) &
                        (self.db.zf_forum.id == self.db.zf_topic.forum_id) &
                        (self.db.zf_topic.parent_flag==True) \
                    ).select(self.db.zf_topic.title,
                        self.db.zf_topic.id,
                        self.db.zf_topic.modifying_date,
                        self.db.zf_topic.content,
                        orderby=~self.db.zf_topic.modifying_date,
                        limitby=(0, max_topics))
            else:
                latest_topics = self.db(
                    (self.db.zf_topic.system_announcement_flag == False) &
                    (self.db.zf_forum.include_latest_topics == True) &
                    (self.db.zf_forum.id == self.db.zf_topic.forum_id) &
                    (self.db.zf_topic.parent_flag == True)).select(
                    self.db.zf_topic.title,
                    self.db.zf_topic.id,
                    self.db.zf_topic.modifying_date,
                    orderby=~self.db.zf_topic.modifying_date,
                    limitby=(0, max_topics))
            if not (latest_topics) and not rss:
                latest_topics = [{'error':'No Topics'}]
        else:
            if not rss:
                latest_topics = [{'error':'No Topics'}]
        return latest_topics

    def setup_notifications(self, subscription_id, subscription_type, now):
        """ subscription_id is the topic.id or forum.id being updated,
        subscription_type is 'T' (Topic) or 'F' (Forum)

        """
        # Grab all users subscribed to this object
        subscribers = self.db(
            (self.db.zf_member_subscriptions.subscription_active == True) &
            (self.db.zf_member_subscriptions.subscription_id == \
             subscription_id) &
            (self.db.zf_member_subscriptions.subscription_type == \
             subscription_type)).select(
            self.db.zf_member_subscriptions.user_id)
        if subscribers:
            for subscriber in subscribers:
                # Only add this subscription notification email request
                # if another one for the same type/object/validity does
                # not exist yet..
                is_subscribed = self.db(
                    (self.db.zf_member_subscriptions_notification.user_id == subscriber.user_id) &
                    (self.db.zf_member_subscriptions_notification.subscription_type == subscription_type) %
                    (self.db.zf_member_subscriptions_notification.subscription_id == subscription_id) &
                    (self.db.zf_member_subscriptions_notification.is_processed == False)).count() > 0
                if not is_subscribed:
                    self.db.zf_member_subscriptions_notification.insert(
                        user_id=subscriber.user_id,
                        subscription_id=subscription_id,
                        subscription_type=subscription_type,
                        creation_date=now,
                        is_processed=False)

    def get_inappropriate_topics_cnt(self):
        return self.db(self.db.zf_topic_inappropriate.read_flag ==\
                       False).count()

    def get_admin_msgs_cnt(self):
        return self.db(self.db.zf_admin_messages.read_flag == False).count()

    def get_my_messages_cnt(self):
        msg_count = self.db(
            (self.db.zf_pm.user_id == self.auth_user.get_user_id()) &
            (self.db.zf_pm.read_flag == False)).count()
        if msg_count > 0:
            rval = '<b style="color:red;">(%s)</b>' % (msg_count)
        else:
            rval = '(%s)' % (msg_count)
        return rval

    def get_image_info(self, data):
        data = str(data)
        size = len(data)
        height = -1
        width = -1
        content_type = ''

        # handle GIFs
        if (size >= 10) and data[:6] in ('GIF87a', 'GIF89a'):
            # Check to see if content_type is correct
            content_type = 'image/gif'
            w, h = struct.unpack("<HH", data[6:10])
            width = int(w)
            height = int(h)

        # See PNG v1.2 spec (http://www.cdrom.com/pub/png/spec/)
        # Bytes 0-7 are below, 4-byte chunk length, then 'IHDR'
        # and finally the 4-byte width, height
        elif ((size >= 24) and (data[:8] == '\211PNG\r\n\032\n') and
            (data[12:16] == 'IHDR')):
            content_type = 'image/png'
            w, h = struct.unpack(">LL", data[16:24])
            width = int(w)
            height = int(h)

        # Maybe this is for an older PNG version.
        elif (size >= 16) and (data[:8] == '\211PNG\r\n\032\n'):
            # Check to see if we have the right content type
            content_type = 'image/png'
            w, h = struct.unpack(">LL", data[8:16])
            width = int(w)
            height = int(h)

        elif (size >= 2) and (data[:2] == '\377\330'):
            content_type = 'image/jpeg'
            jpeg = StringIO(data)
            jpeg.read(2)
            b = jpeg.read(1)
            try:
                while (b and ord(b) != 0xDA):
                    while (ord(b) != 0xFF):
                        b = jpeg.read(1)
                    while (ord(b) == 0xFF):
                        b = jpeg.read(1)
                    if (ord(b) >= 0xC0 and ord(b) <= 0xC3):
                        jpeg.read(3)
                        h, w = struct.unpack(">HH", jpeg.read(4))
                        break
                    else:
                        jpeg.read(int(struct.unpack(">H", jpeg.read(2))[0])-2)
                    b = jpeg.read(1)
                width = int(w)
                height = int(h)
            except:
                pass

        return content_type, width, height

    def gen_pwd(self):
        vowels='aeiou'
        consonants='bcdfghjklmnpqrstvwxyz'
        password=''
        # Some default values should be in the parameters section
        minpairs = 4
        maxpairs = 6

        for x in range(1, random.randint(int(minpairs), int(maxpairs))):
            consonant = consonants[random.randint(1, len(consonants) - 1)]
            if random.choice([1,0]):
                consonant = consonant.upper()
            password = password + consonant
            vowel = vowels[random.randint(1, len(vowels) - 1)]
            if random.choice([1,0]):
                vowel = vowel.upper()
            password = password + vowel

        return password

    def pagination_widget(self, total, start, href, ptype='forum'):
        """ A Pagination Widget """

        # total = total number of results
        # start = zero-based index of first result to be displayed (e.g.
        # page 3 of 5 10-batch pages starts with 20)
        # href  = url + necessary querystring for your results

        # Items per page
        if ptype == 'forum':
            batch = int(self.get_system_property('zfsp_threads_per_page', 15))
        elif ptype == 'topic':
            batch = int(self.get_system_property(
                'zfsp_responses_per_page', 15))
        elif ptype == 'admin_users':
            batch = 50
        else:
            batch = 15

        # current page number. add first to prevent zero division errors.
        current = (start + batch) / batch

        # total number of pages. use ceiling to catch a remainder.
        pages = int(math.ceil(float(total) / float(batch)))

        # Need to modify the widget to return a more user-friendly pagination
        # UI, in the form:
        # [Previous 10] [Previous] 1 2 3 4 5 6 7 [8] 9 10 [Next] [Next 10]
        # If Click on "next 10" then show:
        # [Previous 10] [Previous] 11 12 13 14 15 16 17 [18] 19 20 [Next]
        #     [Next 10]
        # Assuming there are 25 "pages":
        # [Previous 10] [Previous] 21 22 23 24 [25] [Next] [Next 10]
        # Things to consider:
        # - If on page 28 of a 32 page widget, click on "Next 10" must go to
        # the "last" (32) number, not 38
        # - "Previous", "Previous 10", "Next" and "Next 10" must be disabled
        # when appropriate

        PAGE_SET = 10
        sets_to_display = pages / PAGE_SET
        reminder_pages  = pages % PAGE_SET
        if reminder_pages:
            extra_page = 1
        else:
            extra_page = 0
        sets_to_display += extra_page

        real_sets_to_display = total / float(batch)

        if real_sets_to_display > 1:
            html_code = '<div class="breadcrumbs" style="text-align:center;"> '
            '<span i18n:translate="">Pages</span>:\n'

            if start >= (batch * PAGE_SET):
                html_code += '[<a class="breadcrumbs" href="'
                html_code += '%s?start=%s">Previous %s</a>]&nbsp;' % (
                    href, start - (batch * PAGE_SET), PAGE_SET)

            if start - batch >= 0:
                html_code += '[<a class="breadcrumbs" href='
                html_code += '"%s?start=%s" title="">Previous</a>]&nbsp;' % (
                    href, start - batch)

            # Need to find the right "range", for example on a 10-page set of
            # 3-results-per-page for 44 total results,
            # If I click page 6, the range must be 1 2 3 4 5 6 7 8 9 10
            # If I am on page 11, then click on 15, the range must be 11 12
            # 13 14 15 16 17 18 19 20
            # and so on (Note 16 17 18 19 20 must be disabled)
            start_number = 0
            end_number = PAGE_SET
            mid_number = (start / batch) + 1
            while 1:
                if mid_number >= start_number and mid_number <= end_number:
                    break
                start_number = end_number
                end_number = start_number + PAGE_SET

            page = start_number * batch

            this_range = range(1 + start_number, end_number + 1)

            for idx in this_range:
                if idx == (start + batch) / batch:
                    html_code += '<b>%s</b>&nbsp;' % str(idx)
                else:
                    if idx * batch - batch < total:
                        html_code += '<a class="breadcrumbs" href="'
                        html_code += '%s?start=%s" title="">%s</a>&nbsp;' % (
                            href, page, idx)

                page += batch

            if start + batch < total:
                html_code += '[<a class="breadcrumbs" href="'
                html_code += '%s?start=%s" title="">Next</a>]&nbsp;' % (
                    href, start + batch)

            if start + (batch * PAGE_SET) < total:
                tot = start + (batch * PAGE_SET)
                html_code += '[<a class="breadcrumbs" href="'
                html_code += '%s?start=%s" title="">Next %s</a>]&nbsp;' % (
                    href, tot, PAGE_SET)

            html_code += '</div>\n'
        else:
            html_code = ''

        return html_code
